pkg/receiver/gitproviderreceiver/internal/scraper/gitlabscraper/gitlab_scraper.go

```diff
@@ -70,55 +75,58 @@ func (gls *gitlabScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
 
 	graphClient := graphql.NewClient("https://gitlab.com/api/graphql", gls.client)
 
-	subgroups, err := getGroupDescendents(context.Background(), graphClient, gls.cfg.GitLabOrg)
-
-	var projectList []string
+	var projectList []gitlabProject
+	var projectsCursor *string
 
-	if len(subgroups.Group.DescendantGroups.Nodes) > 0 {
-		for _, d := range subgroups.Group.DescendantGroups.Nodes {
-			projectList = append(projectList, d.Name)
-		}
-	}
-
-	// TODO: this only works for top level groups with 1 level of subgroups
-	for _, group := range subgroups.Group.DescendantGroups.Nodes {
-		// full path should be a function that takes in a list of strings and returns a string
-		// concatenated together. This makes it testable & works with multiple subgroups
-		projects, err := getGroupProjects(context.Background(), graphClient, gls.cfg.GitLabOrg+"/"+group.Name)
+	for hasNextPage := true; hasNextPage; {
+		// Get the next page of data
+		projects, err := getAllGroupProjects(context.Background(), graphClient, gls.cfg.GitLabOrg, projectsCursor)
 		if err != nil {
-			gls.logger.Sugar().Errorf("error getting descendant groups: %v", err)
-		}
-		for _, project := range projects.Group.Projects.Nodes {
-			projectList = append(projectList, project.Name)
+			gls.logger.Sugar().Errorf("error: %v", err)
 		}
-	}
 
-	// TODO: Must paginate and update query to get all branches of projects in the list
-	//for _, project := range projectList {
+		if len(projects.Group.Projects.Nodes) == 0 {
+			err = errors.New("no GitLab projects found for the given group/org")
+			gls.logger.Sugar().Error(err)
 
-	//	gls.logger.Sugar().Debugf("project: %v", project)
+			return gls.mb.Emit(), err
+		}
 
-	//	branches, err := getBranchNames(context.Background(), graphClient, project)
+		// Check if there is a next page
+		hasNextPage = projects.Group.Projects.PageInfo.HasNextPage
 
-	//	if err != nil {
-	//		gls.logger.Sugar().Errorf("error getting branches: %v", err)
-	//	}
+		// Set the cursor to the end cursor of the current page
+		projectsCursor = &projects.Group.Projects.PageInfo.EndCursor
 
-	//	var branchList []string
-	//	branchList = append(branchList, branches.Project.Repository.BranchNames...)
+		for _, p := range projects.Group.Projects.Nodes {
+			projectList = append(projectList, gitlabProject{
+				Name:           p.Name,
+				Path:           p.FullPath,
+				CreatedAt:      p.CreatedAt,
+				LastActivityAt: p.LastActivityAt,
+			})
+		}
+	}
 
-	//	gls.logger.Sugar().Debugf("branch: %v", branches.Project.Repository.BranchNames)
+	// A map that maps a project name to an array of branch names.
+	var branchNames map[string][]string = make(map[string][]string, len(projectList))
 
-	//	gls.mb.RecordGitRepositoryBranchCountDataPoint(now, int64(len(branchList)), project)
+	// TODO: Must account for when there are more than 100,000 branch names in a project.
```

Comment:
I'm wondering if we should make a comment + code that says if number of repo's > x number, then emit metric, but don't go and make x number of requests for all the different repos with specific branch information, etc.

Reply:
If I understand you correctly, that's what it's doing now but with `x = 100,000`. It'll get those 100,000 repos and then emit the count. It will get the branch metrics for each of those 100,000 repos though.

Are you saying to have something like `x=10,000` and we get up to 100,000 to report for repo count but only get branch metrics for the 10,000?

Reply:
Yes, a case where it's so high that we don't get the branch info (plus more info) for each of the repos. Ideally though we can sort repos instead of just getting everything.